xilem_core/src/deferred.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{fmt::Display, marker::PhantomData};
+
+use alloc::{boxed::Box, sync::Arc};
+
+use crate::{DynMessage, Message, NoElement, View, ViewId, ViewPathTracker};
+
+/// A `Context` for a [`View`](crate::View) implementation which supports
+/// asynchronous message reporting.
+pub trait AsyncCtx<Message = DynMessage>: ViewPathTracker {
+    /// Get a [`Proxy`] for this context.
+    // TODO: Maybe store the current path within this Proxy?
+    fn proxy(&mut self) -> Arc<dyn RawProxy<Message>>;
+}
+
+/// A handle to a Xilem driver which can be used to queue a message for a View.
+///
+/// These messages are [`crate::DynMessage`]s, which are sent to a view at
+/// a specific path.
+///
+/// This can be used for asynchronous event handling.
+/// For example, to get the result of a `Future` or a channel into
+/// the view, which then will ultimately.
+///
+/// In the Xilem crate, this will wrap an `EventLoopProxy` from Winit.
+///
+/// ## Lifetimes
+///
+/// It is valid for a [`Proxy`] to outlive the [`View`](crate::View) it is associated with.
+pub trait RawProxy<Message = DynMessage>: Send + Sync + 'static {
+    /// Send a `message` to the view at `path` in this driver.
+    ///
+    /// Note that it is only valid to send messages to views which expect
+    /// them, of the type they expect.
+    /// It is expected for [`View`](crate::View)s to panic otherwise, and the routing
+    /// will prefer to send stable.
+    ///
+    /// # Errors
+    ///
+    /// This method may error if the driver is no longer running, and in any other
+    /// cases directly documented on the context which was used to create this proxy.
+    /// It may also fail silently.
+    // TODO: Do we want/need a way to asynchronously report errors back to the caller?
+    //
+    // e.g. an `Option<Arc<dyn FnMut(ProxyError, ProxyMessageId?)>>`?
+    fn send_message(&self, path: Arc<[ViewId]>, message: Message) -> Result<(), ProxyError>;
+}
+
+/// A way to send a message of an expected type to a specific view.
+pub struct MessageProxy<M: Message> {
```

Comment:
That `Message` bound is a little bit unfortunate for xilem_web, as it makes it incompatible with `web_sys` types (mostly relevant for `web_sys::Event` and derivatives).
If we want to make this compatible, something like `RawProxy<ErasedMessage = DynMessage>` would also be necessary for this. But at that point, as mentioned in the final review comment, I think it should rather be a trait (as the impl of `message` doesn't work that easily anymore).